### PR TITLE
Dev temp

### PR DESCRIPTION
**Pull request by MasterVexillen**
## Changelog
- Addition of robust checks for input parameters
- Added AUTOBIN functionality for automatically determining internal binning factor on a per tomogram basis
- Optimisation of slicing algorithm
- Addition of adaptive mode for tomograms with high variability on first pass
- Addition of YAML parameter input support
- Usage of jackknife distance for outlier removal
- Addition of auto-generation of bash scripts for IMOD point2model
- Addition of function for 3D lamella region masks as output
- Splitting of analytics from raw data
- Improvement in on-screen results output
- Addition of GeoLlama task metadata in output STAR file
- Improvement of generic on-screen logging during GeoLlama tasks
- Update of unit tests
- Update of README

### CODE REVIEW

File: GeoLlama/calc_by_slice.py [hunk 1]
```diff
@@ -84,31 +86,32 @@ def autocontrast_slice(image: npt.NDArray[any],
...
     Returns:
     ndarray, ndarray
     """
 
-    std_half = std_window // 2
-    z_range = (max(0, coords[0]-std_half), min(coords[0]+std_window-std_half, volume.shape[0]-1))
-    x_range = (max(0, coords[1]-std_half), min(coords[1]+std_window-std_half, volume.shape[1]-1))
-    y_range = (max(0, coords[2]-std_half), min(coords[2]+std_window-std_half, volume.shape[2]-1))
+    # Create sliding window views
+    sliding_zy = swv(volume, sliding_window_width, axis=1)
```
Comment (elainehoml) [inline GeoLlama/calc_by_slice.py:105]: had no idea this existed, very neat

File: GeoLlama/calc_by_slice.py [hunk 2]
```diff
@@ -280,41 +269,43 @@ def evaluate_slice(slice_coords: list,
     mask_s1 = np.argwhere(view**2 >= thres)
```
Comment (elainehoml) [inline GeoLlama/calc_by_slice.py:269]: why are the pixel values being squared?

File: GeoLlama/calc_by_slice.py [hunk 3]
```diff
@@ -280,41 +269,43 @@ def evaluate_slice(slice_coords: list,
...
 
     # Skip slice if no pixels masked
-    # if len(mask_s1) < pt_thres:
-    #     return
-    # centroid_s1 = 0.5*(mask_s1.max(axis=0) + mask_s1.min(axis=0))
     centroid_s1 = mask_s1.mean(axis=0)
 
     # Step 2: Use Mahalanobis distance to remove potential outliers
```
Comment (elainehoml) [inline GeoLlama/calc_by_slice.py:274]: jackknife*

File: GeoLlama/calc_by_slice.py [hunk 4]
```diff
@@ -280,41 +269,43 @@ def evaluate_slice(slice_coords: list,
...
-    # Skip slice if no pixels masked
-    # if len(mask_s2) < pt_thres:
-    #     return
+    conf_limit = t.interval(confidence=0.999, df=2,
+                            loc=jackknife_dist.mean(), scale=sem(jackknife_dist))[1]
+    mask_s2 = np.squeeze(mask_s1[np.argwhere(jackknife_dist<=conf_limit)], axis=1)
 
     # Step 3: Use distance-based clustering to find sample region
-    clusters = DBSCAN(eps=15, min_samples=15).fit_predict(mask_s2)
+    clusters = DBSCAN(eps=15, min_samples=350).fit_predict(mask_s2)
```
Comment (elainehoml) [inline GeoLlama/calc_by_slice.py:287]: Would the user ever want to tweak the `eps` and `min_samples` for clustering? Were these default values are set by trial and error?

File: GeoLlama/calc_by_slice.py [hunk 5]
```diff
@@ -280,41 +269,43 @@ def evaluate_slice(slice_coords: list,
...
     mask_s3_args = np.argwhere(clusters==mode(clusters, keepdims=True).mode)
     mask_s3 = mask_s2[mask_s3_args.flatten()]
 
     # Skip slice if no pixels masked
```
Comment (elainehoml) [inline GeoLlama/calc_by_slice.py:291]: do the lines below actually skip the slice? I'm a little confused by this comment

File: GeoLlama/calc_by_slice.py [hunk 6]
```diff
@@ -330,73 +321,23 @@ def evaluate_slice(slice_coords: list,
     cell_vecs = eigenvecs * rectangle_dims.reshape((2, 1))
 
     # Centralise lamella centroid to middle of slice along long axis
-    lamella_centre = centroid_s3
-
-    extrema = np.array(view.shape) - 1
-    ref_pt1 = lamella_centre + cell_vecs[1]
-    ref_pt2 = lamella_centre - cell_vecs[1]
+    lamella_to_slice_dist = 200 * max( np.abs(lamella_centre/np.array(view.shape) - 0.5) )
```
Comment (elainehoml) [inline GeoLlama/calc_by_slice.py:324]: where did 200 come from?

File: GeoLlama/config.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+) -> objects.Config:
+    """
+    Objectifying user provided input as a Config object
+
+    Args:
+    autocontrast (bool) : Apply autocontrast to slices prior to evaluation
+    adaptive (bool) : Use adaptive sampling for slice evaluation
+    bandpass (bool) : Apply bandpass filter to tomograms prior to evaluation
+    user_path (str) : Path to folder holding all tomograms in batch mode
+    pixel_size (float) : Tomogram pixel size in nm
```
Comment (elainehoml) [inline GeoLlama/config.py:131]: Might want to be quite explicit here - if the mdocs say my pixel size is 1 A, and I've reconstructed my tomograms with bin 8, is the `pixel_size` for Geollama supposed to be 0.1 nm or 0.8 nm?

File: GeoLlama/main.py [hunk 8]
```diff
@@ -16,121 +16,267 @@
...
+        data_path: Annotated[
+            typing.Optional[str],
+            typer.Option("-p", "--path",
+                         help="Path to folder holding all tomograms. (NB. Direct path to individual tomogram file not supported.)"),
+        ] = None,
+        pixel_size_nm: Annotated[
+            typing.Optional[float],
+            typer.Option(
+                "-s", "--pixel_size",
+                help="Tomogram pixel size in nm."),
```
Comment (elainehoml) [inline GeoLlama/main.py:90]: see comment in `config.py` on what `pixel_size` means

File: README.md [hunk 9]
```diff
@@ -18,6 +18,24 @@ If pytest is not available, you can also run tests from the package root folder
 python3 -m unittest discover
 ```
 
+## Quickstarter
+To use GeoLlama, first create a YAML configuration file by running the command
+```
+geollama generate-config [--output-path PATH]
+```
+The configuration will be generated in the current working directory with default parameters if the `--output-path` flag is not provided.
```
Comment (elainehoml) [inline README.md:26]: slightly confusing wording - if i specify the output path will the default parameters still be in the yaml?

**Review by elainehoml — COMMENTED** [review]
Looks good - loads of work that you've done on this since the last time! Just a few comments here and there mostly on clarifying comments/documentation.

File: GeoLlama/calc_by_slice.py [hunk 10]
```diff
@@ -249,28 +250,16 @@ def evaluate_slice(slice_coords: list,
     Function for creation and evaluation of one 2D slice from tomogram given coordinates of intersection point
 
     Args:
-    slice_coords (list) : list containing coordinates of intersection point for image slicing
     volume (ndarray) : tomogram for evaluation
```
[the remark this answers is not in the document]
Comment (MasterVexillen) [reply GeoLlama/calc_by_slice.py:253]: It is assumed to be ZXY. But the problem with the coordinate system we are now using is David Mastronarde's invention, and when we view 2D images in python, especially using `plt.imshow`, the convention is (row, column) which means your the X axis in your XY plane/view becomes the row axis and the Y axis becomes the column axis. But we would want the other way round, so I added in a transpose.

File: GeoLlama/calc_by_slice.py (hunk 2 above)
Comment (MasterVexillen) [reply GeoLlama/calc_by_slice.py:269]: I could've just used the magic number 1.12 times the threshold (I've mentioned it in the manuscript) - I'll change it in the code.

File: GeoLlama/calc_by_slice.py (hunk 6 above)
Comment (MasterVexillen) [reply GeoLlama/calc_by_slice.py:324]: I'm calculating the percentage of shift in terms of the _semi-major_ axis of the 2D image, not the full major axis, so the 200 actually comes from 100 / 0.5.

File: GeoLlama/config.py (hunk 7 above)
Comment (MasterVexillen) [reply GeoLlama/config.py:131]: It is the input pixel size, so 0.1 in your example. Will change in the code.

File: README.md (hunk 9 above)
Comment (MasterVexillen) [reply README.md:26]: Oh, I meant 

> A YAML configuration file will be generated with default parameters. If the `--output-path` flag is not provided, the file will be generated at `./config.yaml` by default.

Will be amended.

File: GeoLlama/calc_by_slice.py (hunk 4 above)
Comment (MasterVexillen) [reply GeoLlama/calc_by_slice.py:287]: They won't - they are set by trial and error

**Pull request merged**